### PR TITLE
Fix for 4250

### PR DESCRIPTION
**Pull request by ChamaraPhilipsuom**
Integrated two if clauses to give claims for valid token

### CODE REVIEW

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/CacheClearingUserOperationListener.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+                    AuthorizationGrantCacheKey cacheKey = new AuthorizationGrantCacheKey(accessToken);
+                    AuthorizationGrantCacheEntry cacheEntry = (AuthorizationGrantCacheEntry) AuthorizationGrantCache
+                            .getInstance().getValueFromCacheByToken(cacheKey);
+                    if (cacheEntry != null) {
+                        AuthorizationGrantCache.getInstance().clearCacheEntryByToken(cacheKey);
+                    }
+                }
+            }
+        } catch (IdentityOAuth2Exception e) {
+            String errorMsg = "Error occurred while retrieving access tokens issued for user : " + userName;
```
Comment (cdwijayarathna) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/CacheClearingUserOperationListener.java]: Is it necessary to declare a new variable here?

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOathEventListener.java [hunk 2]
```diff
@@ -76,8 +79,8 @@ public boolean doPreDeleteUser(java.lang.String username,
 
     @Override
     public boolean doPostSetUserClaimValues(String userName, Map<String, String> claims, String profileName,
```
Comment (ChamaraPhilipsuom) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOathEventListener.java]: move to presetuserClaimValues. Override doPreSetUSerClaimValue also.

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java [hunk 3]
```diff
@@ -81,6 +81,8 @@
 
     private static final String IDN_OAUTH2_ACCESS_TOKEN = "IDN_OAUTH2_ACCESS_TOKEN";
 
+    private static final String ACTIVE_STATE = "ACTIVE";
```
Comment (johannnallathamby) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java]: Use OAuthConstants.TokenStates.TOKEN_STATE_ACTIVE

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java [hunk 4]
```diff
@@ -973,6 +975,143 @@ public void revokeToken(String tokenId, String userId) throws IdentityOAuth2Exce
     }
 
     /**
+     *
+     * @param tenantId
+     * @param userName
+     * @param userId
+     * @throws IdentityOAuth2Exception
+     */
+    public Set<String> getAccessTokenListForUserByTenant(String tenantId, String userName, String userId) throws IdentityOAuth2Exception {
```
Comment (johannnallathamby) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java]: Pass only AuthenticatedUser

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java [hunk 5]
```diff
@@ -973,6 +975,143 @@ public void revokeToken(String tokenId, String userId) throws IdentityOAuth2Exce
...
+    }
+
+    /**
+     *
+     * @param tenantId
+     * @param userName
+     * @param userId
+     * @throws IdentityOAuth2Exception
+     */
+    public Set<String> getAuthorizationCodeListForUserByTenant(String tenantId, String userName, String userId) throws IdentityOAuth2Exception {
```
Comment (johannnallathamby) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java]: Pass only AuthenticatedUser

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuthApplicationMgtListener.java [hunk 6]
```diff
@@ -66,6 +74,7 @@ public boolean doPostUpdateApplication(ServiceProvider serviceProvider, String t
 
         addClientSecret(serviceProvider);
         updateAuthApplication(serviceProvider);
+        removeAccessTokensAndAuthCodeFromCache(serviceProvider,tenantDomain,userName);
```
Comment (johannnallathamby) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuthApplicationMgtListener.java]: formatting issues

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOathEventListener.java (hunk 2 above)
Comment (johannnallathamby) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOathEventListener.java]: Let's implement this in pre listener. Also need to implement doPreSetUserClaimValue()

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOathEventListener.java [hunk 7]
```diff
@@ -180,4 +183,47 @@ private boolean revokeTokens(String username, UserStoreManager userStoreManager)
         }
         return true;
     }
+
+    private void removeAccessTokenFromCache(String userName, UserStoreManager userStoreManager) throws
```
Comment (johannnallathamby) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOathEventListener.java]: Rename as removeTokensFromCache()

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java [hunk 8]
```diff
@@ -973,6 +975,143 @@ public void revokeToken(String tokenId, String userId) throws IdentityOAuth2Exce
...
+            IdentityDatabaseUtil.closeAllConnections(connection, null, ps);
+        }
+        return authorizationCodeList;
+    }
+
+    public Set<String> getActiveTokenListForConsumerKey(String consumerKey) throws IdentityOAuth2Exception {
+        Connection connection = IdentityDatabaseUtil.getDBConnection();
+        PreparedStatement ps = null;
+        ResultSet rs = null;
+        Set<String> accessTokenList = new HashSet<>();
```
Comment (johannnallathamby) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java]: Rename as accessTokenSet

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java [hunk 9]
```diff
@@ -973,6 +975,143 @@ public void revokeToken(String tokenId, String userId) throws IdentityOAuth2Exce
...
+            IdentityDatabaseUtil.rollBack(connection);
+            throw new IdentityOAuth2Exception("Error occurred while revoking Access Token with user Name : " +
+                    userName + " tenant ID : " + tenantId, e);
+        }  finally {
+            IdentityDatabaseUtil.closeAllConnections(connection, null, ps);
+        }
+        return authorizationCodeList;
+    }
+
+    public Set<String> getActiveTokenListForConsumerKey(String consumerKey) throws IdentityOAuth2Exception {
```
Comment (johannnallathamby) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java]: Rename to getActiveTokensForConsumerKey

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java [hunk 10]
```diff
@@ -973,6 +975,143 @@ public void revokeToken(String tokenId, String userId) throws IdentityOAuth2Exce
...
+            throw new IdentityOAuth2Exception("Error occurred while getting access tokens from acces token table for " +
+                    "the application with consumer key : " + consumerKey, e);
+        } finally {
+            IdentityDatabaseUtil.closeAllConnections(connection, null, ps);
+        }
+        return accessTokenList;
+    }
+
+    public Set<String> getAuthorizationCodeListForConsumerKey(String consumerKey) throws IdentityOAuth2Exception {
+        Connection connection = IdentityDatabaseUtil.getDBConnection();
```
Comment (johannnallathamby) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java]: Rename to getAuthorizationCodesForConsumerKey

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java [hunk 11]
```diff
@@ -973,6 +975,143 @@ public void revokeToken(String tokenId, String userId) throws IdentityOAuth2Exce
...
+        PreparedStatement ps = null;
+        ResultSet rs = null;
+        Set<String> authorizationCodeList = new HashSet<>();
```
Comment (johannnallathamby) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java]: Rename to authorizationCodes

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java [hunk 12]
```diff
@@ -973,6 +975,143 @@ public void revokeToken(String tokenId, String userId) throws IdentityOAuth2Exce
...
+     * @param userName
+     * @param userId
+     * @throws IdentityOAuth2Exception
+     */
+    public Set<String> getAuthorizationCodeListForUserByTenant(String tenantId, String userName, String userId) throws IdentityOAuth2Exception {
+
+        Connection connection = IdentityDatabaseUtil.getDBConnection();
+        PreparedStatement ps = null;
+        ResultSet rs = null;
+        Set<String> authorizationCodeList = new HashSet<>();
```
Comment (johannnallathamby) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java]: Rename to authorizationCodes

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java [hunk 13]
```diff
@@ -973,6 +975,143 @@ public void revokeToken(String tokenId, String userId) throws IdentityOAuth2Exce
...
+     * @param userId
+     * @throws IdentityOAuth2Exception
+     */
+    public Set<String> getAccessTokenListForUserByTenant(String tenantId, String userName, String userId) throws IdentityOAuth2Exception {
+
+        String accessTokenStoreTable = OAuthConstants.ACCESS_TOKEN_STORE_TABLE;
+        Connection connection = IdentityDatabaseUtil.getDBConnection();
+        PreparedStatement ps = null;
+        ResultSet rs = null;
+        Set<String> accessTokenList = new HashSet<>();
```
Comment (johannnallathamby) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java]: Rename to accessTokens

**Pull request merged**